metricflow-semantics/metricflow_semantics/model/semantics/linkable_element_set.py

```diff
@@ -290,3 +288,104 @@ def only_unique_path_keys(self) -> LinkableElementSet:
                 if len(set(linkable_metrics)) <= 1
             },
         )
+
+    @property
+    @override
+    def derived_from_semantic_models(self) -> Sequence[SemanticModelReference]:
+        semantic_model_references: Set[SemanticModelReference] = set()
+        for linkable_dimensions in self.path_key_to_linkable_dimensions.values():
+            for linkable_dimension in linkable_dimensions:
+                semantic_model_references.update(linkable_dimension.derived_from_semantic_models)
+        for linkable_entities in self.path_key_to_linkable_entities.values():
+            for linkable_entity in linkable_entities:
+                semantic_model_references.update(linkable_entity.derived_from_semantic_models)
+        for linkable_metrics in self.path_key_to_linkable_metrics.values():
+            for linkable_metric in linkable_metrics:
+                semantic_model_references.update(linkable_metric.derived_from_semantic_models)
+
+        return sorted(semantic_model_references, key=lambda reference: reference.semantic_model_name)
+
+    @property
+    def spec_count(self) -> int:
+        """If this is mapped to spec objects, the number of specs that would be produced."""
+        return (
+            len(self.path_key_to_linkable_dimensions.keys())
+            + len(self.path_key_to_linkable_entities.keys())
+            + len(self.path_key_to_linkable_metrics.keys())
+        )
+
+    @property
+    def specs(self) -> Sequence[LinkableInstanceSpec]:
+        """Converts the items in a `LinkableElementSet` to their corresponding spec objects."""
+        specs: List[LinkableInstanceSpec] = []
+
+        for path_key in (
+            tuple(self.path_key_to_linkable_dimensions.keys())
+            + tuple(self.path_key_to_linkable_entities.keys())
+            + tuple(self.path_key_to_linkable_metrics.keys())
+        ):
+            specs.append(self._path_key_to_spec(path_key))
+
+        return specs
+
+    def _path_key_to_spec(self, path_key: ElementPathKey) -> LinkableInstanceSpec:
+        if path_key.element_type is LinkableElementType.DIMENSION:
+            return DimensionSpec(
+                element_name=path_key.element_name,
+                entity_links=path_key.entity_links,
+            )
+        elif path_key.element_type is LinkableElementType.TIME_DIMENSION:
+            assert path_key.time_granularity is not None
```

Comment:
nit: error message, maybe something like "type refinement, should have been checked in dataclass validation"

Reply:
Updated.